File: app/src/main/java/com/kotlinblog/actors/view/MainViewModel.java
package com.kotlinblog.actors.view;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.kotlinblog.actors.App;
import com.kotlinblog.actors.data.Actor;
import com.kotlinblog.actors.data.ActorsList;
import com.kotlinblog.actors.data.Api;
import com.kotlinblog.actors.utils.SingleLiveEvent;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
import timber.log.Timber;

public class MainViewModel extends ViewModel {

    @Inject Retrofit mRetrofit;
    private MutableLiveData<ActorsList> mActors = new MutableLiveData<>();
    private SingleLiveEvent<String> mConnectionError = new SingleLiveEvent<>();
    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    public MainViewModel() {
        App.getComponent().inject(this);
    }

    void fetchActors() {
        Api service = mRetrofit.create(Api.class);
        mCompositeDisposable.add(service.getActors()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        response -> {
                            Timber.d(response.toString());
                            mActors.postValue(response);
                        },
                        error -> {
                            Timber.e(error.getMessage());
                            mConnectionError.postValue(error.getMessage());
                        }
                )
        );
    }

    public MutableLiveData<ActorsList> getActors() {
        return mActors;
    }

    public SingleLiveEvent<String> getConnectionError() {
        return mConnectionError;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (!mCompositeDisposable.isDisposed()) {
            mCompositeDisposable.dispose();
        }
    }

    /**
     * Adapter callbacks
     */
    Actor getActorAt(int position) {
        if (position < getActorsListSize() && mActors.getValue() != null) {
            return mActors
                    .getValue()
                    .getActors()
                    .get(position);
        } else {
            return null;
        }
    }

    int getActorsListSize() {
        if (mActors.getValue() != null) {
            return mActors
                    .getValue()
                    .getActors()
                    .size();
        } else {
            return 0;
        }
    }
}

File: app/src/main/java/com/kotlinblog/actors/view/MainViewModel.kt
package com.kotlinblog.actors.view

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.kotlinblog.actors.App
import com.kotlinblog.actors.data.Actor
import com.kotlinblog.actors.data.ActorsList
import com.kotlinblog.actors.data.Api
import com.kotlinblog.actors.utils.SingleLiveEvent
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import timber.log.Timber
import javax.inject.Inject

class MainViewModel : ViewModel() {

    @Inject lateinit var mRetrofit: Retrofit

    private var mActors: MutableLiveData<ActorsList>
    val actors: LiveData<ActorsList> get() = mActors

    private var mConnectionError: SingleLiveEvent<String>
    val connectionError: SingleLiveEvent<String> get() = mConnectionError

    private val mCompositeDisposable = CompositeDisposable()

    init {
        App.getComponent().inject(this)
        mActors = MutableLiveData()
        mConnectionError = SingleLiveEvent()
    }

    fun fetchActors() {
        val service = mRetrofit.create<Api>(Api::class.java)
        mCompositeDisposable.add(service.getActors()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    mActors.postValue(response)
                }, {
                    error ->
                    Timber.e(error.message)
                    connectionError.postValue(error.message)
                }))
    }

    override fun onCleared() {
        super.onCleared()
        if (!mCompositeDisposable.isDisposed) {
            mCompositeDisposable.dispose()
        }
    }

    /**
     * Adapter callbacks
     */
    fun getActorAt(position: Int): Actor? {
        return if (position < getMealListSize()) {
            actors.value?.actors?.get(position)
        } else {
            null
        }
    }

    fun getMealListSize(): Int {
        actors.value?.actors?.let{
            return it.size
        }
        return 0
    }
}